keps/sig-scheduling/20200116-coscheduling-plugin-based-on-scheduler-framework.md

```diff
@@ -0,0 +1,217 @@
+``` 
+
+### Coscheduling
+```go
+// Coscheduling is a plugin that implements the mechanism of gang scheduling.
+type Coscheduling struct {
+    FrameworkHandle     	framework.FrameworkHandle
+    PodLister               corelisters.PodLister
+    // Key is the name of PodGroup.
+    PodGroupInfos       	map[string]PodGroupInfo
+    // Name of the last scheduled podgroup
+    LastPodGroup        	string
+}
+ 
+type PodGroupInfo struct {
+    // LastFailureTimestamp stores the timestamp of last scheduling failure.
+    LastFailureTimestamp 	time.Time
+    UID           			types.UID
+    MinAvailable  		int
+    Name                                 String
+}
+```
+
+1.  `PodGroupInfo` is initialized the first time the pod belongs to the PodGroup is encountered, and LastFailureTimestamp is updated every time the PodGroup fails to schedule.
+2.  `LastPodGroup` records which `PodGroup` the last scheduled pod belongs to.
+3.  `UID` is the unique identification value used to distinguish different podgroups.
+ 
+
+### Extension points
+
+#### QueueSort
+In order to make the pods which belongs to the same `PodGroup` to be scheduled together as much as possible, implement a strategy in `QueueSort` phase. 
+
+```go
+  func  Less(podA *PodInfo, podB *PodInfo) bool
+```
+1. Trying to order by pod priority (i.e. .spec.priorityValue), pod with higher priority is scheduled ahead of other pod with lower priority. When priorities are the same, they are operated according to the following process.
+   
+2. When podA and podB are both regularPods (we will check it by their labels), it follows the same logic of default in-tree [PrioritySort](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/queuesort/priority_sort.go#L41-L45) plugin.
+   
+3. When podA is regularPod, podB is pgPod, trying to order by podA's `Timestamp` (the time pod added to the scheduling queue) and podB’s `LastFailureTimestamp` (we get the PodGroupInfo from the cache). Pod with earlier timestamp is scheduled ahead of other pod.   
+   
+4. When podA and podB are both pgPods.
+   1. Trying to order by the `LastFailureTimestamp` of the podGroup. Pod with earlier timestamp is scheduled ahead of other pod.
+   
+   2. If `LastFailureTimestampA` is equal to `LastFailureTimestampB`, trying to order by the `UID` of `PodGroup`,`Pod` with lexicographically greater `UID` is scheduled ahead of other pod. (The purpose is to distinguish different `PodGroup` with the same `LastFailureTimestamp` and to keep the pods of the same `PodGroup` together)
+
+**Note1**: There are different `LastFailureTimestamp` (even if they are the same, the UID will be different). So when the pods enter the queue, the pods that belongs to the same PodGroup will be together.
+
+#### Pre-Filter
+1. `PreFilter` validates that if the total number of pods belonging to the same `PodGroup` is less than `minAvailable`. If so, the scheduling process will be interrupted directly.
```

Comment:
Is this trying to make sure all pods that belong to a podgroup are already in the queues?

Also, what is "interrupted"? I assume this means that it would return `UnschedulableAndUnresolvable`? What happens with `LastFailureTimestamp`?

Reply:
Have we given up on the idea of only allowing one podgroup to be scheduled at a time? This design might still lead to deadlocks if more than one pod group already has minpods in the queues.

Reply:
sorting the pods by their group id in the queue tries to approximate scheduling one pod group at a time, but this approach indeed does not offer any guarantees.

Reply:
> This design might still lead to deadlocks if more than one pod group already has minpods in the queues.

I think so. So I'm questioning why it only allows ONE on-going PodGroup. See my comment below.